### ISSUE TITLE
support packing scripts build defined entry-point console_script/gui_script

### ISSUE THREAD

**Issue by htgoebel** [pyinstaller/pyinstaller#305]
**Original date**: 2010/12/30

If a project uses setuptools `console_script` or `gui_script` entry-points for generating scripts, pyinstaller is not able to package this scripts properly.

An example for this can be found in [pdfposter](http://pdfposter.origo.ethz.ch/) ([direct link to setup.py](http://svn.origo.ethz.ch/wsvn/pdfposter/trunk/setup.py)). The actual script `pdfposter` is generated when the package is installed.
### Analysis of the problem

The generated scripts do not import any module, but only contain a call like

```
load_entry_point('pdftools.pdfposter==0.5.0', 'console_scripts', 'pdfposter')()
```

Thus pyinstaller is not able to find the required package.
### Solution outline

`Analysis()` needs to get another parameter `hidden_imports`. This will allow adding the required module in a .spec file.

Optionally: if #32 (integration into distutils) is finished, the setup-command could could access the list of entry-points and add the "hidden_imports" automatically.
### Workaround - Recipe

https://github.com/pyinstaller/pyinstaller/wiki/Recipe-Setuptools-Entry-Point

**Comment by tartley** [comment]
A comment above contains a link to the recipe that is now broken. The correct link is now:
https://github.com/pyinstaller/pyinstaller/wiki/Recipe-Setuptools-Entry-Point

**Comment by matysek** [comment]
@tartley thanks I updated the description with the link.

**Comment by nchammas** [comment]
I have this issue as well. My project does not have a script that you can run by calling `python script.py`. Instead, it registers a `console_scripts` entry point which places an executable on the user's `PATH`.

Can PyInstaller package itself? It looks like PyInstaller [uses `console_scripts` entry points as well](https://github.com/pyinstaller/pyinstaller/blob/0c7b9021bfa0ce1a0a3d6d9396b095291c72830c/setup.py#L108-L117). Another popular package which would have the same issue is [httpie](https://github.com/jkbrzt/httpie/blob/c6690e0182ac924e30a7c4800686f26a0aa7e5ad/setup.py#L82-L86).

I guess there are two work-arounds available today:

1. Use the [recipe mentioned earlier here](https://github.com/pyinstaller/pyinstaller/wiki/Recipe-Setuptools-Entry-Point), and
2. Modify your project to additionally allow execution via the usual `python script.py`

I think in the long-term it is important to support `console_scripts` directly so that the user story for PyInstaller becomes this: "If it works with `python setup.py sdist` and `python setup.py bdist_wheel`, then it will work with PyInstaller--no extra work required." That would be ideal for users.

**Comment by ckochenower** [comment]
Note that if you use the [recipe](https://github.com/pyinstaller/pyinstaller/wiki/Recipe-Setuptools-Entry-Point), "workpath" is lowercase. I made a small tweak to hiddenimports kwarg and the packages = hiddenimports or [] assignment to avoid a NonType exception when trying to iterate over hiddenimports:

    # My tweaked version of Entrypoint
    def Entrypoint(dist, group, name, 
               scripts=None, pathex=None, hiddenimports=[],
               hookspath=None, excludes=None, runtime_hooks=None):
    import pkg_resources

    # get toplevel packages of distribution from metadata
    def get_toplevel(dist):
        distribution = pkg_resources.get_distribution(dist)
        if distribution.has_metadata('top_level.txt'):
            return list(distribution.get_metadata('top_level.txt').split())
        else:
            return []

    packages = hiddenimports
    for distribution in hiddenimports:
        packages += get_toplevel(distribution)

    scripts = scripts or []
    pathex = pathex or []
    # get the entry point
    ep = pkg_resources.get_entry_info(dist, group, name)
    # insert path of the egg at the verify front of the search path
    pathex = [ep.dist.location] + pathex
    # script name must not be a valid module name to avoid name clashes on import
    script_path = os.path.join(workpath, name + '-script.py')
    #script_path = os.path.join(BUILDPATH, name + '-script.py')
    print "creating script for entry point", dist, group, name
    with open(script_path, 'w') as fh:
        fh.write("import {0}\\\\n".format(ep.module_name))
        fh.write("{0}.{1}()\\\\n".format(ep.module_name, '.'.join(ep.attrs)))
        for package in packages:
            fh.write("import {0}\\\\n".format(package))

    return Analysis([script_path] + scripts, pathex, hiddenimports, hookspath, excludes, runtime_hooks)

**Comment by kawing-chiu** [comment]
+1 for this. Using entry_points is really common practice nowadays. Hope that this issue will not be postponed again...

**Comment by justinmayer** [comment]
We are running into this entry point problem as well, and we haven't had any success with the linked workaround recipe. [I posted the details](http://stackoverflow.com/questions/39599940/pyinstaller-relative-imports-and-entry-points) and would greatly appreciate any suggestions for successfully resolving this problem.

**Comment by xoviat** [comment]
Is there a particular reason that the EntryPoint class is not in the codebase?

**Comment by xoviat** [comment]
I am currently working on a resolution to this problem using code from cx_Freeze (which has some strengths but cannot compile as many programs as pyinstaller). However, the problem is that the code that I am using is licensed under the Python Software Foundation License but Pyinstaller is licensed under GPL, so it may turn out that my code will not be able to be merged into Pyinstaller.

**Comment by xoviat** [comment]
Also, I'm not going to wait on this to proceed on my end, but it would be very nice not to have to generate a spec file to call the main build process. Perhaps allow a function to be passed that accepts the environment as arguments?

**Comment by htgoebel** [comment]
> Is there a particular reason that the EntryPoint class is not in the codebase?

Simply that nobody send a pull-request for that :-)

**Comment by nateshmbhat** [comment]
In reference to issue #1086 , have been experiencing problems with subliminal package while compiling . I have used the Entry point recipe . But unfortunately its not working right .

Here is the error during compiling : - 

```
C:\Users\Natesh\Desktop\subtitle trying>pyinstaller Mynewspecfile.spec
self.modulename = PyInstaller.__main__  ; __name__ = pkg_resources
49 INFO: PyInstaller: 3.3.dev0+gb899bfbd
49 INFO: Python: 3.5.2
49 INFO: Platform: Windows-8.1-6.3.9600-SP0
49 INFO: UPX is not available.
Traceback (most recent call last):
  File "C:\Program Files\Python35\Scripts\pyinstaller-script.py", line 9, in <module>
    load_entry_point('PyInstaller==3.3.dev0+gb899bfbd', 'console_scripts', 'pyinstaller')()
  File "c:\program files\python35\lib\site-packages\PyInstaller\__main__.py", line 92, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "c:\program files\python35\lib\site-packages\PyInstaller\__main__.py", line 46, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "c:\program files\python35\lib\site-packages\PyInstaller\building\build_main.py", line 790, in main
    build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'))
  File "c:\program files\python35\lib\site-packages\PyInstaller\building\build_main.py", line 736, in build
    exec(text, spec_namespace)
  File "<string>", line 45, in <module>
  File "<string>", line 26, in Entrypoint
AttributeError: 'NoneType' object has no attribute 'dist'
```

**Comment by htgoebel** [comment]
@nateshmbhat This ticket is about *implementing* EntryPoint as a standard function in PyInstaller. You text is about an *problem* with a very specific package. Please do not pulte this ticket but open a new one. Thanks.

**Comment by Legorooj** [comment]
Stale

**Comment by delijati** [comment]
Could someone provide a example repo.

**Comment by tartley** [comment]
pdfposter is mentioned in this issue description, but the link is bad. A working link is:
https://gitlab.com/pdftools/pdfposter/-/blob/develop/setup.py
(could someone with privs update the issue description)

**Comment by ericoporto** [comment]
Is the recipe here still relevant? (https://github.com/pyinstaller/pyinstaller/wiki/Recipe-Setuptools-Entry-Point)
